1. Importing Libraries

In [37]:
# libraries 

import numpy as np, pandas as pd, re, os

In [38]:
# setting display options 
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

2. Reading the data 

In [39]:
PROJECT_DIR = "/Users/abhisheksaurav/Desktop/real_estate_project/"
DATA_DIR = "Data Collection/Data/City/chandigarh/"

In [40]:
def get_data(name):
    file_name = "{}.csv".format(name)
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    return pd.read_csv(file_path)

In [41]:
df = get_data('house')

In [42]:
df.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69

3. Preliminary Analysis

3.1 Check Data Types

In [43]:
df.dtypes

property_name      object
link               object
society            object
price              object
rate               object
area               object
areaWithType       object
bedRoom            object
bathroom           object
balcony            object
additionalRoom     object
address            object
noOfFloor          object
facing             object
agePossession      object
nearbyLocations    object
description        object
furnishDetails     object
features           object
rating             object
property_id        object
dtype: object

In [44]:
df.shape

(1095, 21)

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    1095 non-null   object
 1   link             1095 non-null   object
 2   society          471 non-null    object
 3   price            968 non-null    object
 4   rate             1054 non-null   object
 5   area             1095 non-null   object
 6   areaWithType     987 non-null    object
 7   bedRoom          987 non-null    object
 8   bathroom         987 non-null    object
 9   balcony          987 non-null    object
 10  additionalRoom   589 non-null    object
 11  address          1031 non-null   object
 12  noOfFloor        967 non-null    object
 13  facing           674 non-null    object
 14  agePossession    987 non-null    object
 15  nearbyLocations  913 non-null    object
 16  description      1036 non-null   object
 17  furnishDetails   743 non-null    

3.2 Checking for Duplicate Values 

In [46]:
(
    df
    .duplicated()
    .sum()
)

np.int64(13)

In [47]:
(
    df 
    .loc[df.duplicated(keep = False)]
    .sort_values(['property_id', 'property_name', 'address'])
)

property_name  \
58       4 Bedroom House for sale in Sector 91 Gurgaon   
830      4 Bedroom House for sale in Sector 91 Gurgaon   
55       3 Bedroom House for sale in Sector 48 Gurgaon   
827      3 Bedroom House for sale in Sector 48 Gurgaon   
53      4 Bedroom House for sale in Sector 112 Gurgaon   
825     4 Bedroom House for sale in Sector 112 Gurgaon   
48      4 Bedroom House for sale in Sector 70A Gurgaon   
820     4 Bedroom House for sale in Sector 70A Gurgaon   
59       4 Bedroom House for sale in Sector 86 Gurgaon   
831      4 Bedroom House for sale in Sector 86 Gurgaon   
56       4 Bedroom House for sale in Sector 48 Gurgaon   
828      4 Bedroom House for sale in Sector 48 Gurgaon   
57       4 Bedroom House for sale in Sector 55 Gurgaon   
829      4 Bedroom House for sale in Sector 55 Gurgaon   
54         6 Bedroom House for sale in Uppals Southend   
826        6 Bedroom House for sale in Uppals Southend   
52          5 Bedroom House for sale in Greenwood City   
824         5 Bedroom House for sale in Greenwood City   
50             3 Bedroom House for sale in Palam Vihar   
822            3 Bedroom House for sale in Palam Vihar   
792           12 Bedroom House for sale in DLF Phase 1   
871           12 Bedroom House for sale in DLF Phase 1   
51             8 Bedroom House for sale in Palam Vihar   
823            8 Bedroom House for sale in Palam Vihar   
49   4 Bedroom House for sale in C Block Sushant Lo...   
821  4 Bedroom House for sale in C Block Sushant Lo...   

                                                  link  \
58   https://www.99acres.com/4-bhk-bedroom-independ...   
830  https://www.99acres.com/4-bhk-bedroom-independ...   
55   https://www.99acres.com/3-bhk-bedroom-independ...   
827  https://www.99acres.com/3-bhk-bedroom-independ...   
53   https://www.99acres.com/4-bhk-bedroom-independ...   
825  https://www.99acres.com/4-bhk-bedroom-independ...   
48   https://www.99acres.com/4-bhk-bedroom-independ...   
820  https://www.99acres.com/4-bhk-bedroom-independ...   
59   https://www.99acres.com/4-bhk-bedroom-independ...   
831  https://www.99acres.com/4-bhk-bedroom-independ...   
56   https://www.99acres.com/4-bhk-bedroom-independ...   
828  https://www.99acres.com/4-bhk-bedroom-independ...   
57   https://www.99acres.com/4-bhk-bedroom-independ...   
829  https://www.99acres.com/4-bhk-bedroom-independ...   
54   https://www.99acres.com/6-bhk-bedroom-independ...   
826  https://www.99acres.com/6-bhk-bedroom-independ...   
52   https://www.99acres.com/5-bhk-bedroom-independ...   
824  https://www.99acres.com/5-bhk-bedroom-independ...   
50   https://www.99acres.com/3-bhk-bedroom-independ...   
822  https://www.99acres.com/3-bhk-bedroom-independ...   
792  https://www.99acres.com/12-bhk-bedroom-indepen...   
871  https://www.99acres.com/12-bhk-bedroom-indepen...   
51   https://www.99acres.com/8-bhk-bedroom-independ...   
823  https://www.99acres.com/8-bhk-bedroom-independ...   
49   https://www.99acres.com/4-bhk-bedroom-independ...   
821  https://www.99acres.com/4-bhk-bedroom-independ...   

                         society       price             rate  \
58     DLF New Town Heights3.7 ★   2.5 Crore   ₹ 7,058/sq.ft.   
830    DLF New Town Heights3.7 ★   2.5 Crore   ₹ 7,058/sq.ft.   
55            Vipul Tatvam Villa  7.25 Crore  ₹ 27,971/sq.ft.   
827           Vipul Tatvam Villa  7.25 Crore  ₹ 27,971/sq.ft.   
53      Experion Windchants3.8 ★  7.99 Crore  ₹ 12,519/sq.ft.   
825     Experion Windchants3.8 ★  7.99 Crore  ₹ 12,519/sq.ft.   
48              Bptp Visionnaire   5.4 Crore  ₹ 20,690/sq.ft.   
820             Bptp Visionnaire   5.4 Crore  ₹ 20,690/sq.ft.   
59   DLF New Town Heights 24.0 ★   2.5 Crore   ₹ 7,058/sq.ft.   
831  DLF New Town Heights 24.0 ★   2.5 Crore   ₹ 7,058/sq.ft.   
56            Vipul Tatvam Villa   8.4 Crore  ₹ 21,000/sq.ft.   
828           Vipul Tatvam Villa   8.4 Crore  ₹ 21,000/sq.ft.   
57                           NaN  3.55 Crore  ₹ 24,500/sq.ft. 

Observation: 
1. There are 1095 rows and 21 columns
2. All columns are "object" types which is wrong. 
3. We will follow the same steps as we did on flats.csv file to clean it 
4. There are 13 duplicate rows which we need to drop
5. we will drop `property_id`, `link` as these columns doesn't contribute anything to our analysis. 

4. Detailed Analysis

In [48]:
df.head()

property_name  \
0    5 Bedroom House for sale in Sector 70A Gurgaon   
1  5 Bedroom House for sale in Sector 21A Faridabad   
2  10 Bedroom House for sale in Sushant Lok Phase 1   
3    21 Bedroom House for sale in Sector 54 Gurgaon   
4  12 Bedroom House for sale in Sushant Lok Phase 1   

                                                link           society  \
0  https://www.99acres.com/5-bhk-bedroom-independ...  Bptp Visionnaire   
1  https://www.99acres.com/5-bhk-bedroom-independ...               NaN   
2  https://www.99acres.com/10-bhk-bedroom-indepen...               NaN   
3  https://www.99acres.com/21-bhk-bedroom-indepen...               NaN   
4  https://www.99acres.com/12-bhk-bedroom-indepen...               NaN   

        price              rate                   area  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating property_id  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...   K70037724  
1                                                NaN   E69288322  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...   F69536898  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...   R69

4.1 Society

In [62]:
(
    df
    .society
    .value_counts()
)

society
Vipul Tatvam Villa                                   27
International City by SOBHA Phase 2                  26
Emaar MGF Marbella                                   26
International City by Sobha Phase 1                  23
Unitech Uniworld Resorts                             14
DLF City Plots                                       11
DLF City Plots Phase 2                               11
Unitech Espace                                       11
Bptp Visionnaire                                      9
Eldeco Mansionz                                       9
Uppal Southend4.0 ★                                   9
Not Applicable                                        7
Experion Windchants3.8 ★                              6
Ansals Palam Vihar                                    6
Sushant Lok 1 Builder Floors                          6
DLF New Town Heights 24.0 ★                           5
DLF City Phase 1                                      5
Vatika India Next4.1 ★                  

5. Cleaning Operations

In [60]:
def clean_data(df):
    return (
        df
        .drop_duplicates()
        .drop(columns = ['property_id', 'link'])
        .rename(columns = {"area": "price_per_sq_ft"})
        
    )

In [61]:
house_cleaned = clean_data(df)
house_cleaned.head()

property_name           society  \
0    5 Bedroom House for sale in Sector 70A Gurgaon  Bptp Visionnaire   
1  5 Bedroom House for sale in Sector 21A Faridabad               NaN   
2  10 Bedroom House for sale in Sushant Lok Phase 1               NaN   
3    21 Bedroom House for sale in Sector 54 Gurgaon               NaN   
4  12 Bedroom House for sale in Sushant Lok Phase 1               NaN   

        price              rate        price_per_sq_ft  \
0  5.25 Crore   ₹ 20,115/sq.ft.  (242 sq.m.) Plot Area   
1   5.7 Crore  ₹ 105,751/sq.ft.   (50 sq.m.) Plot Area   
2   2.1 Crore   ₹ 38,251/sq.ft.   (51 sq.m.) Plot Area   
3     5 Crore   ₹ 43,066/sq.ft.  (108 sq.m.) Plot Area   
4     3 Crore   ₹ 53,763/sq.ft.   (52 sq.m.) Plot Area   

                  areaWithType      bedRoom      bathroom       balcony  \
0  Plot area 290(242.48 sq.m.)   5 Bedrooms   4 Bathrooms  3+ Balconies   
1   Plot area 539(50.07 sq.m.)   5 Bedrooms   4 Bathrooms   2 Balconies   
2       Plot area 61(51 sq.m.)  10 Bedrooms  10 Bathrooms  3+ Balconies   
3  Plot area 129(107.86 sq.m.)  21 Bedrooms  21 Bathrooms  3+ Balconies   
4    Plot area 62(51.84 sq.m.)  12 Bedrooms  12 Bathrooms  3+ Balconies   

                       additionalRoom  \
0                        Servant Room   
1  Store Room,Pooja Room,Servant Room   
2                        Servant Room   
3                        Servant Room   
4                              Others   

                                       address noOfFloor      facing  \
0    29b, Sector 70A Gurgaon, Gurgaon, Haryana  3 Floors  North-East   
1       Sector 21A Faridabad, Gurgaon, Haryana  2 Floors         NaN   
2        Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   
3          Sector 54 Gurgaon, Gurgaon, Haryana  5 Floors       North   
4  1228, Sushant Lok Phase 1, Gurgaon, Haryana  5 Floors        West   

      agePossession                                    nearbyLocations  \
0   0 to 1 Year Old  ['Good Earth City Center 2', 'Kunskapsskolan I...   
1  5 to 10 Year Old                                                NaN   
2   0 to 1 Year Old  ['Sector 42-43 metro station', 'Huda city cent...   
3   1 to 5 Year Old  ['Sector 53-54 metro station', 'Sector 54 chow...   
4   Within 6 months  ['Sector 42-43 metro station', 'Huda city cent...   

                                         description  \
0  Do you wish to buy an independent house in sec...   
1  Hi, we have an independent house/villa availab...   
2  Monthly rental income is rs1,40,000/- Best opt...   
3  129 sq yd plot size. 5 floors built up , fully...   
4  Best for investment purpose. Fully furnished b...   

                                      furnishDetails  \
0  ['1 Wardrobe', '1 Fan', '1 Exhaust Fan', '1 Ge...   
1  ['1 Water Purifier', '5 Fan', '1 Exhaust Fan',...   
2  ['10 Bed', '3 Fan', '10 Geyser', '2 Light', 'N...   
3  ['1 Water Purifier', '21 Fan', '1 Fridge', '1 ...   
4  ['1 Water Purifier', '1 Fridge', '12 Fan', '1 ...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
1  ['Private Garden / Terrace', 'Park', 'Visitor ...   
2  ['Maintenance Staff', 'Water Storage', 'Visito...   
3  ['Feng Shui / Vaastu Compliant', 'Private Gard...   
4  ['Maintenance Staff', 'Water Storage', 'Visito...   

                                              rating  
0  ['Environment5 out of 5', 'Lifestyle4 out of 5...  
1                                                NaN  
2  ['Environment5 out of 5', 'Lifestyle5 out of 5...  
3  ['Environment4 out of 5', 'Lifestyle5 out of 5...  
4  ['Environment5 out of 5', 'Lifestyle5 out of 5...